content/chat/rooms/occupancy.textile

```diff
@@ -0,0 +1,99 @@
+---
+title: Occupancy
+meta_description: "Use occupancy to see how many users are in a room."
+product: chat
+languages:
+  - javascript
+---
+
+Occupancy enables you to view the number of users in a room. This feature can be used to display room counts or highlight popular, or trending chat rooms.
+
+<aside data-type='note'>
+<p>Occupancy needs to be "enabled":/chat/rooms#create when creating or retrieving a room.</p>
+</aside>
+
+h2(#subscribe). Subscribe to room occupancy
+
+Subscribe to a room's occupancy by registering a listener. Occupancy events are emitted whenever the membership of a room changes. Use the "@occupancy.subscribe()@":https://sdk.ably.com/builds/ably-labs/ably-chat-js/main/typedoc/interfaces/Occupancy.html#subscribe method in a room to receive updates:
+
+```[javascript]
+const {unsubscribe} = room.occupancy.subscribe((event) => {
+  console.log(event);
+});
+```
+
+h3(#payload). Occupancy payload
+
+The following is an example payload of an occupancy event:
+
+```[json]
+{
+  "connections": 103;
+  "presenceMembers": 95;
+}
+```
+
+The following are the properties of an occupancy event payload:
+
+|_. Property |_. Description |_. Type |
+| connections | The number of connections in the room. | Number |
+| presenceMembers | The number of users entered into the "presence set":/chat/rooms/presence of the room. | Number |
+
+Two different metrics are returned so that you can differentiate between anonymous and identified users. A user must be identified with a @clientId@ in order to register their "presence":/chat/rooms/presence in a room. Therefore @presenceMembers@ is a count of identified users in the room, and @connections@ returns the number of anonymous and identified users in the room.
```

Comment:
We always talked about chat as if it were in a logged-in setting, where the ably customer's servers would somehow identify all clients.

What happens when a non-logged-in user goes to a page that has an ably chat on it? We have to call out that their server should be able to issue "anonymous" tokens somehow with perhaps random client IDs, and provide some recommended readonly permissions (eg be in occupancy, see occupancy, don't show in presence but perhaps see presence of others, see messages but don't post messages, etc)? cc @AndyTWF and @splindsay-92 since this isn't something we cared or talked about.

Reply:
Have been discussing this with Andy and it's something I'm thinking about for the permission and role management milestone (tbd). For now, do you think it would be confusing to say what I've suggested above? Do you think we'd have to provide users with a lot more guidance for this concept to be useful? @vladvelici

Reply:
It's just this sentence that grabbed my attention

> Two different metrics are returned so that you can differentiate between anonymous and identified users

In my mind those features are just different but don't tell us anything about anonymous or identified users. One is a user list (presence) and one is numbers of online users/etc (occupancy).

The _with account_ or _without account_ issue I'd say it's out of scope for us but we can include a sentence/paragraph to recommend random unique clientIds that are consistent between page refreshes (important if non-logged-in are allowed to send messages, so their own messages still show as theirs; presence can get weird if you change clientId on refresh as well since you become someone else on each refresh)?

Reply:
So if we go for `clientId` being mandatory then the difference between the occupancy metrics only boils down to those that have entered into presence, and those that have not.

Reply:
yup